# Initial Data Understanding

## Loading Data and Libraries

In [208]:
import pandas as pd
import numpy as np
import sys
import os

import seaborn as sns
import matplotlib.pyplot as plt

sys.path.append('../src/')
pd.set_option('display.max_columns', None)

%matplotlib inline
%load_ext autoreload
%autoreload 2

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


## Análise preliminar dos dados de Quimioterapia e Radioterapia para Câncer de Mama
Neste documento serão tratadas as primeiras identificações de padrão e possíveis incongruências - de acordo com o julgamento do grupo - em relação aos dados fornecidos para Quimioterapia e Radioterapia de pacientes atendidos para câncer de mama.

In [209]:
kwargs = {'encoding':'latin-1', 'error_bad_lines':False, 'sep':',', 'low_memory':False, 'index_col':0, 'skiprows':0}

df_mama_quim = pd.read_csv(os.path.join('../data/', 'raw', 'Linfomas Quimioterapia SIA-SUS.csv'), **kwargs)
df_mama_rad = pd.read_csv(os.path.join('../data/', 'raw', 'Linfomas Radioterapia SIA-SUS.csv'), **kwargs)

In [210]:
import constants

df_mama_quim.rename(columns=constants.LAYOUT_APAC, inplace=True)
df_mama_rad.rename(columns=constants.LAYOUT_APAC, inplace=True)

### Amostras de dados
Um primeiro panorama de formatos e disposição de ambos datasets.

In [211]:
df_mama_quim.sample(5)

dt_process tp_gestao  cod_tp_gestao  cod_est_cnes       num_apac  \
4       201401        EP         120000       2001586  1213200185967   
25      201401        EP         120000       2001586  1213200175561   
27      201401        EP         120000       2001586  1213200175528   
28      201401        EP         120000       2001586  1213200177266   
32      201401        EP         120000       2001586  1213200186275   

    dt_atend_paciente  proc_princ_apac  vl_tot_apac_aprov  cod_uf_mun  tp_est  \
4              201401        304060011            1258.64      120040       5   
25             201401        304060011            1258.64      120040       5   
27             201401        304060011            1258.64      120040       5   
28             201401        304060011            1258.64      120040       5   
32             201401        304070017            1700.00      120040       5   

    tip_prest est_man_ind  cnpj_est_exe  cnpj_mante          num_csn  \
4          40           I  6.360294e+13         NaN  ‚{}{{}ƒ{{„„€   
25         40           I  6.360294e+13         NaN  ƒ„ƒ{{|{|‚|ƒ€   
27         40           I  6.360294e+13         NaN  ƒ„ƒ{{{{€|}{|}   
28         40           I  6.360294e+13         NaN  ƒ„ƒ{{{{{}~}€   
32         40           I  6.360294e+13         NaN  ƒ„ƒ{{{ƒ{ƒƒ{‚‚   

    cod_idade  num_idade sexo  raca_cor  cod_uf_mun_paciente  nac_paciente  \
4           4         20    M         3               120040            10   
25          4         46    M         3               120040            10   
27          4         20    M         3               120040            10   
28          4         61    M         3               120001            10   
32          4         18    F         3               120040            10   

    cep_paciente  uf_res_dif_paciente  mun_res_dif_paciente  dt_ini_val  \
4       69900970                    0                     1    20131105   
25      69900970                    0                     1    20131105   
27      69900970                    0                     1    20131105   
28      69945000                    0                     1    20131105   
32      69900970                    0                     1    20131128   

    dt_fim_val  tp_atend_apac  ind_apac  mot_sai_perm  ind_obito  ind_encerr  \
4     20140131              3         2            21          0           0   
25    20140131              3         2            21          0           0   
27    20140131              3         2            21          0           0   
28    20140131              3         2            21          0           0   
32    20140131              3         2            21          0           0   

    ind_perm  ind_alta  ind_transf  dt_ocorr cod_org_emi  car_atend  \
4          1         0           0       NaN  E120000001          1   
25         1         0           0       NaN  E120000001          1   
27         1         0           0       NaN  E120000001          1   
28         1         0           0       NaN  E120000001          1   
32         1         0           0       NaN  E120000001          1   

     num_apac_ant cod_soli_cnes     dt_soli   dt_auto cid_caus_assoc  \
4   0000000000000       2001586  20131105.0  20131105           0000   
25  0000000000000       0000000  20131105.0  20131105           0000   
27  0000000000000       0000000  20131105.0  20131105           0000   
28  0000000000000       0000000  20131105.0  20131105           0000   
32  0000000000000       2001586  20131128.0  20131128           0000   

                  cid_princ cid_sec etnia aq_cid_10 aq_linfo_reg_inva  \
4   C81- Linfoma de Hodgkin    0000   NaN       NaN                 3   
25  C81- Linfoma de Hodgkin    0000   NaN       NaN                 3   
27  C81- Linfoma de Hodgkin    0000   NaN       NaN                 3   
28  C81- Linfoma de Hodgkin    0000   NaN       NaN                 3   
32  C81- Linfoma de Hodgk

In [213]:
df_mama_rad.sample(5)

dt_process tp_gestao  cod_tp_gestao  cod_est_cnes       num_apac  \
8       201401        EP         120000       2001586  1214200029119   
14      201402        EP         120000       2001586  1214200045201   
21      201402        EP         120000       2001586  1214200029119   
29      201403        EP         120000       2001586  1214200045201   
76      201405        EP         120000       2001586  1214200084944   

    dt_atend_paciente  proc_princ_apac  vl_tot_apac_aprov  cod_uf_mun  tp_est  \
8              201401        304010090                614      120040       5   
14             201402        304010090               2098      120040       5   
21             201402        304010090                630      120040       5   
29             201403        304010090               2550      120040       5   
76             201405        304010090                810      120040       5   

    tip_prest est_man_ind    cnpj_est_exe  cnpj_mante          num_csn  \
8          40           I  63602940000170         NaN  ƒ„ƒ{{~„€ƒ„ƒ€ƒ   
14         40           I  63602940000170         NaN  ƒ„ƒ{{}„€~~ƒ{}   
21         40           I  63602940000170         NaN  ƒ„ƒ{{~„€ƒ„ƒ€ƒ   
29         40           I  63602940000170         NaN  ƒ„ƒ{{}„€~~ƒ{}   
76         40           I  63602940000170         NaN  }{{€{{{|}{{{~   

    cod_idade  num_idade sexo  raca_cor  cod_uf_mun_paciente  nac_paciente  \
8           4          8    M         3               120050            10   
14          4         75    M         3               120040            10   
21          4          8    M         3               120050            10   
29          4         75    M         3               120040            10   
76          4         13    M         3               120020            10   

    cep_paciente  uf_res_dif_paciente  mun_res_dif_paciente  dt_ini_val  \
8       69940000                    0                     1    20140123   
14      69900970                    0                     1    20140212   
21      69940000                    0                     1    20140123   
29      69900970                    0                     1    20140212   
76      69980000                    0                     1    20140318   

    dt_fim_val  tp_atend_apac  ind_apac  mot_sai_perm  ind_obito  ind_encerr  \
8     20140228              4         1            21          0           0   
14    20140430              4         1            21          0           0   
21    20140228              4         2            15          0           0   
29    20140430              4         2            15          0           0   
76    20140531              4         2            15          0           0   

    ind_perm  ind_alta  ind_transf    dt_ocorr cod_org_emi  car_atend  \
8          1         0           0         NaN  E120000001          1   
14         1         0           0         NaN  E120000001          1   
21         0         1           0  20140228.0  E120000001          1   
29         0         1           0  20140331.0  E120000001          1   
76         0         1           0  20140531.0  E120000001          1   

    num_apac_ant  cod_soli_cnes     dt_soli     dt_auto cid_caus_assoc  \
8            0.0      2001586.0  20140123.0  20140123.0           0000   
14           0.0      2001586.0  20140212.0  20140212.0           0000   
21           0.0      2001586.0  20140123.0  20140123.0           0000   
29           0.0      2001586.0  20140212.0  20140212.0           0000   
76           0.0      2001586.0  20140318.0  20140318.0           0000   

                  cid_princ cid_sec  etnia  ar_smrd ar_cid_10  \
8   C81- Linfoma de Hodgkin    0000    NaN      NaN       NaN   
14  C81- Linfoma de Hodgkin    0000    NaN      NaN       NaN   
21  C81- Linfoma de Hodgkin    0000    NaN      NaN       NaN   
29  C81- Linfoma de Hodgkin    0000    NaN      NaN       NaN   
76  C81- Linfoma de Hodgkin  

### Geração de reports
Mesmo com análises exaustivas de frações dos dados, um overview total - de correlações, missing values et al - é imprescindível. Utilizando `pandas_profiling`, geraremos então relatórios que servirão de insumo do cenário macro desses dados.

In [40]:
from pandas_profiling import ProfileReport

ProfileReport(df_mama_quim).to_file('mama_quimioterapia_report.html')
ProfileReport(df_mama_rad).to_file('mama_radioterapia_report.html')

### Missing values

Devemos nos atentar ao valor informativo das colunas - porcentagens muito altas de missing values são indicativos de inaplicabilidade ou simplesmente não preenchimento da informação.

Primeiramente, nos dados de quimioterapia:

In [214]:
1 - df_mama_quim.count().sort_values().head(20) / len(df_mama_quim)

etnia                0.999716
aq_cidini3           0.973344
aq_cidini2           0.937684
aq_dt_ini_3_trat     0.927430
dt_ocorr             0.901655
aq_dt_ini_2_trat     0.887407
aq_cidini1           0.835039
aq_dt_ini_1_trat     0.771918
aq_cid_10            0.670408
aq_esu_p2            0.472046
cnpj_mante           0.271705
num_apac_ant         0.095159
dt_auto              0.075321
aq_estadi            0.041121
dt_soli              0.039926
cod_soli_cnes        0.003130
aq_grahis            0.000527
cnpj_est_exe         0.000035
aq_linfo_reg_inva    0.000000
aq_dt_inden_pato     0.000000
dtype: float64

A porcentagem de valores faltantes em `etnia` é certamente alarmante, uma coluna praticamente 100% nula. Vejamos seus poucos não-nulos

In [215]:
len(df_mama_quim[~df_mama_quim.etnia.isnull()]), len(df_mama_quim)

(90, 316965)

Seguidamente, nos dados de radioterapia:

In [216]:
1 - df_mama_rad.count().sort_values().head(20) / len(df_mama_rad)

ar_smrd             1.000000
etnia               0.999046
ar_cidini3          0.976004
ar_cid_topo_3       0.947691
ar_cidini2          0.943122
ar_dt_ini_3_trat    0.940562
ar_dt_fim_3         0.937952
ar_dt_ini_3         0.937952
ar_dt_ini_2_trat    0.896436
ar_cid_topo_2       0.839408
ar_dt_fim_2         0.830572
ar_dt_ini_2         0.830572
ar_cidini1          0.717871
ar_dt_ini_1_trat    0.659036
ar_num_ins_3        0.562801
dt_ocorr            0.561697
ar_num_ins_2        0.506928
ar_cid_10           0.464056
cnpj_mante          0.359689
ar_trat_anter       0.284137
dtype: float64

O cenário de 99.9%+ nulos observados em `etnia` se repetem também aqui, porém a `ar_smrd` - que inclusive não possui descrição no dicionário correlato - se apresenta com absolutos 100%. Iremos portanto descartar esta coluna.

In [217]:
df_mama_rad = df_mama_rad.drop(columns=['ar_smrd'])

### Valores constantes

De modo similar a missing values, colunas com valores constantes também não apresentam valor significativo. A partir do relatório, notamos essa característica no **Tipo de Atendimento APAC** para ambos sets.

In [218]:
df_mama_quim.tp_atend_apac.unique()

array([3], dtype=int64)

In [219]:
df_mama_rad.tp_atend_apac.unique()

array([4], dtype=int64)

Portanto, serão desconsideradas como colunas.

In [220]:
df_mama_quim = df_mama_quim.drop(columns=['tp_atend_apac'])
df_mama_rad = df_mama_rad.drop(columns=['tp_atend_apac'])

### Agrupamento de datas
Diversas colunas em ambos sets são marcações de tempo, que podem ajudar a inferir a duração dum tratamento.

Como primeiro passo para tratá-las, devemos convertê-las ao formato `datetime`.

In [221]:
dtcols = [c for c in df_mama_quim.columns if c.startswith('dt_')]
df_mama_quim[dtcols].head(10)

dt_process  dt_atend_paciente  dt_ini_val  dt_fim_val  dt_ocorr  \
4       201401             201401    20131105    20140131       NaN   
25      201401             201401    20131105    20140131       NaN   
27      201401             201401    20131105    20140131       NaN   
28      201401             201401    20131105    20140131       NaN   
32      201401             201401    20131128    20140131       NaN   
34      201401             201401    20131205    20140228       NaN   
47      201402             201402    20140204    20140430       NaN   
52      201402             201402    20140207    20140430       NaN   
76      201402             201402    20131205    20140228       NaN   
77      201402             201402    20140218    20140430       NaN   

       dt_soli   dt_auto  
4   20131105.0  20131105  
25  20131105.0  20131105  
27  20131105.0  20131105  
28  20131105.0  20131105  
32  20131128.0  20131128  
34  20131205.0  20131205  
47  20140204.0  20140204  
52  20140207.0  20140207  
76  20131205.0  20131205  
77  20140218.0  20140218

In [225]:
df_mama_quim[dtcols].dtypes

dt_process             int64
dt_atend_paciente      int64
dt_ini_val             int64
dt_fim_val             int64
dt_ocorr             float64
dt_soli              float64
dt_auto               object
dtype: object

As duas primeiras colunas - mês de processamento e atendimento ao paciente - formatam-se em ano-mês, portanto terão conversão diferenciada

In [237]:
ymth_dt = dtcols[0:2]
to_year_month = lambda df: df.apply(pd.to_datetime, format='%Y%M').apply(lambda date: date.to_period('M'))

In [238]:
df_mama_quim[ymth_dt] = df_mama_quim[ymth_dt].apply(to_year_month)
df_mama_rad[ymth_dt] = df_mama_rad[ymth_dt].apply(to_year_month)

df_mama_quim[ymth_dt].head(5)

dt_process dt_atend_paciente
4     2014-01           2014-01
25    2014-01           2014-01
27    2014-01           2014-01
28    2014-01           2014-01
32    2014-01           2014-01

Nota-se que `dt_soli` (data de solicitação) e `dt_ocorr` (data de ocorrência) são colunas `float`, portanto contando com casas decimais que não fazem sentido para o formato `Y-M-d`.

In [271]:
ymd_dt = dtcols[2:-1]
# Retirada de casas decimais e conversão para datetime
str_to_date = lambda date: np.nan if pd.isnull(date) or len(date) != 8 else pd.to_datetime(date, format='%Y%m%d')

float_to_ymd = lambda df: df.fillna(0).astype(int).astype(str).replace('0', np.nan).apply(str_to_date)

In [275]:
df_mama_quim[ymd_dt] = df_mama_quim[ymd_dt].apply(float_to_ymd)
df_mama_rad[ymd_dt] = df_mama_rad[ymd_dt].apply(float_to_ymd)

df_mama_quim[ymd_dt].head(5)

dt_ini_val dt_fim_val dt_ocorr    dt_soli
4  2013-11-05 2014-01-31      NaT 2013-11-05
25 2013-11-05 2014-01-31      NaT 2013-11-05
27 2013-11-05 2014-01-31      NaT 2013-11-05
28 2013-11-05 2014-01-31      NaT 2013-11-05
32 2013-11-28 2014-01-31      NaT 2013-11-28